Opaque Pointer workstream definition (#42)
qir-alliance/.github: workstreams/Opaque_Pointers_Workstream.md

# Workstream: Opaque Pointer Support

## Motivation & Benefits

> Support for typed pointers in LLVM is being removed. More information about the

Reviewer: ```suggestion
Support for typed pointers in LLVM has been removed. More information about the
```

motivation for this change can be found in LLVM forums. For example, the LLVM
documentation on [opaque pointers](https://llvm.org/docs/OpaquePointers.html) as
well as [this
presentation](https://llvm.org/devmtg/2022-04-03/slides/keynote.Opaque.Pointers.Are.Coming.pdf)
give a brief overview.

The QIR specification currently makes use of typed pointers to represent qubit
values and result values. While in LLVM 15, support for these types can still be

> enabled, an update to LLVM 16 requires us to change this representation.

Reviewer: These versions are off by 1 possibly? There is no more typed pointers in version 17 and all future versions of LLVM, I believe. Since typed pointers were removed, there is no backwards compatibility by definition. Said another way, an LLVM front end user that requires typed pointers will be frozen at LLVM version 16.

The goal of this workstream is hence to update the QIR specification to support
newer versions of LLVM, and to document what LLVM versions can be used to
support each version of the QIR specification.

## Requirements

Generally, LLVM maintains very good [backwards
compatibility](https://llvm.org/docs/DeveloperPolicy.html#ir-backwards-compatibility);

> according to the LLVM 18 docs, it is possible to load even bitcode generate by

Reviewer 2: ```suggestion
according to the LLVM 18 docs, it is possible to load even bitcode generated by
```

an LLVM version as early as 3.0. However, it is not possible in general to
generate bitcode that can be loaded with an older version of LLVM than the one
used to emit it. This means that while *QIR consumers*, i.e. backend providers,
can update to the latest LLVM version without sacrificing support for prior QIR

> versions, whereas *QIR producers*, i.e. language frontends, cannot use LLVM

Reviewer 2: ```suggestion
versions, *QIR producers*, i.e. language frontends, cannot use LLVM
```

versions that are newer than the one required by the versions of the QIR
specification they support. This can pose challenges for maintaining a

> state-of-the art frontend.

Reviewer: ```suggestion
state-of-the-art front end.
```
(Front end, the noun, is two words.)

As part of this workstream, we hence specifically propose to update the required

> LLVM version for the next version of the QIR specification to LLVM 16.

Reviewer 3: Looks like LLVM 17 will be the 'latest' release by the time this workstream starts (release is planned for Sept 19th). Would it make sense to always target 'latest' in any 'move LLVM forward' type workstreams to maximize time before we have to migrate again?

Author: updated to llvm 17

## Dependencies & Related Projects

There are no related projects within the QIR Alliance at this time. However,
there are a number of threads in the LLVM community by projects with similar
needs. In particular, the considerations discussed in [this
thread](https://discourse.llvm.org/t/rfc-better-support-for-typed-pointers-in-an-opaque-pointer-world/63339)
and the follow up
[here](https://discourse.llvm.org/t/rfc-adding-opaque-types-to-llvm-ir/65326)
also apply to the QIR ecosystem.

See also the section on [Open Questions](#open-questions) regarding possible
alternatives to consider.

A related topic to consider when discussing options for this workstream is
whether the new representation would also allow to support qudits in the future.

## Deliverable(s) & Expected Outcome

The expected outcome of the workstream is a pull request to update the 0.1
version of the QIR specification with the necessary changes to be able to emit

> the full QIR and all its profile using LLVM 16.

Reviewer 4: The QIR spec update as a result of this work will almost certainly make the new version not backwards compatible with the current specification. Suggest adding migration procedure for current QIR specification users, both emitters of QIR and QIR consumers.

Reviewer 4: Another idea for expected outcome - make modification in a way that it is compatible with older LLVM versions, ideally going back to 10 or even lower. 

This LLVM incompatibility issue can be addressed in a way that makes QIR dependent on opaque pointers added recently or it could be addressed in a way that does not. The former binds QIR specification tighter with the new LLVM features. The latter makes it less LLVM version dependent.

Reviewer 2: Good way of stating it - I would hope that backwards compatibility doesn't become the dominant factor in selecting a course of action but it's wise to note that the working group has a say in just how incompatible the change will be. I don't feel strongly enough about the issue (i.e., support for LLVM < 15) to say that should be a mandate to the working group (as an expected outcome).

Author: I don't think we should mandate it as a prerequisite for the workstream. But it seems fair to ask for a migration plan, so that should ensure the solution is mindful for making the changes in a way that is minimally disruptive.

> the full QIR and all its profile using LLVM 16.

Reviewer: ```suggestion
the full QIR and all its profiles using LLVM 16.
```
I'd suggest version 17 here as the opaque pointer issue is acutely obvious there. Typed pointers, as used in the QIR document, do not exist.

## Future Work (Out of Scope)

> Migration plan for shared components?

Reviewer 3: I don't think we need a 'plan' as such, but certainly understanding the cost and sequencing of migration (e.g. backends first, etc.), and the challenges involved in moving the ecosystem (e.g. if one backend provider can't migrate, is any frontend that wants to support them blocked from migrating?), will be important for success.

Reviewer 4: I had the opposite reaction to this suggesting to make it part of the work stream. See my comment above. New QIR revision will be incompatible with the current one. We need a process for dealing with migration defined to help the current QIR user base adopt new QIR specification version.

Reviewer 2: If I read "shared components" as the other repositories under the alliance umbrella, then it would make sense to assess the consequences of migration by doing it firsthand. It also seems like part of the deliverable here would be a migration plan from spec 0.1 -> 0.2

Author: Ok, let's have this be part of this workstream then.
